File: results/2026-04-07_225702/tasks/12-pr-label-assigner/default-opus/generated-code/label_assigner.py
"""
PR Label Assigner — assigns labels to a PR based on changed file paths.

Uses configurable path-to-label mapping with glob patterns, multiple labels
per file, and priority-based conflict resolution.

Priority logic: when multiple rules match the SAME file, only labels from
the highest-priority (lowest number) matching rules are kept for that file.
Rules at the same priority level all apply. Labels are accumulated across
all files to produce the final set.
"""

from __future__ import annotations

import fnmatch
from dataclasses import dataclass


@dataclass(frozen=True)
class LabelRule:
    """A mapping from a glob pattern to a label, with an optional priority.

    Lower priority numbers win when rules conflict on the same file.
    Default priority is 5.
    """
    pattern: str
    label: str
    priority: int = 5

    def __post_init__(self):
        if not self.pattern:
            raise ValueError(f"Invalid pattern: pattern must be non-empty")
        if not self.label:
            raise ValueError(f"Invalid label: label must be non-empty")
        if self.priority < 0:
            raise ValueError(f"Priority must be a positive integer, got {self.priority}")

    def matches(self, path: str) -> bool:
        """Check if a file path matches this rule's glob pattern.

        Supports ** for recursive directory matching and * for single-segment wildcards.
        For basename-only patterns (no /), matches against just the filename.
        """
        # For patterns without a directory separator, match against both
        # the full path and just the basename (last component)
        if "/" not in self.pattern:
            # e.g. "*.test.*" should match "src/deep/foo.test.js"
            basename = path.rsplit("/", 1)[-1] if "/" in path else path
            return fnmatch.fnmatch(basename, self.pattern)

        # For directory patterns, use fnmatch with the full path
        # fnmatch doesn't natively handle ** so we translate it
        return fnmatch.fnmatch(path, self.pattern)


def assign_labels(files: list[str], rules: list[LabelRule]) -> set[str]:
    """Assign labels to a set of changed files based on the given rules.

    For each file, finds all matching rules, then keeps only those at the
    highest priority (lowest number). The final label set is the union of
    labels chosen across all files.

    Args:
        files: list of changed file paths (relative to repo root)
        rules: list of LabelRule instances

    Returns:
        set of label strings to apply to the PR
    """
    if not files or not rules:
        return set()

    all_labels: set[str] = set()

    for path in files:
        # Collect all matching rules for this file
        matches = [r for r in rules if r.matches(path)]
        if not matches:
            continue

        # Find the best (lowest) priority among matches
        best_priority = min(r.priority for r in matches)

        # Keep only labels from rules at the best priority level
        for r in matches:
            if r.priority == best_priority:
                all_labels.add(r.label)

    return all_labels


def load_rules(config: dict) -> list[LabelRule]:
    """Load label rules from a configuration dictionary.

    Expected format:
        {"rules": [{"pattern": "docs/**", "label": "documentation", "priority": 5}, ...]}

    Args:
        config: dict with a "rules" key containing a list of rule dicts

    Returns:
        list of LabelRule instances

    Raises:
        ValueError: if config is missing required keys or has invalid entries
    """
    if "rules" not in config:
        raise ValueError("Configuration must contain a 'rules' key")

    rules = []
    for i, entry in enumerate(config["rules"]):
        if "pattern" not in entry or "label" not in entry:
            raise ValueError(
                f"Rule at index {i} must have 'pattern' and 'label' keys"
            )
        rules.append(LabelRule(
            pattern=entry["pattern"],
            label=entry["label"],
            priority=entry.get("priority", 5),
        ))

    return rules


# --- CLI entry point for quick demo ---

if __name__ == "__main__":
    # Example configuration
    config = {
        "rules": [
            {"pattern": "docs/**", "label": "documentation", "priority": 5},
            {"pattern": "src/api/**", "label": "api", "priority": 1},
            {"pattern": "src/**", "label": "source", "priority": 10},
            {"pattern": "*.test.*", "label": "tests", "priority": 3},
            {"pattern": "*.md", "label": "markdown", "priority": 8},
            {"pattern": ".github/**", "label": "ci", "priority": 2},
        ]
    }

    # Mock PR changed files
    mock_files = [
        "docs/getting-started.md",
        "src/api/users.py",
        "src/api/users.test.py",
        "src/utils/helpers.py",
        ".github/workflows/ci.yml",
        "README.md",
    ]

    rules = load_rules(config)
    labels = assign_labels(mock_files, rules)

    print("Changed files:")
    for f in mock_files:
        file_matches = [r for r in rules if r.matches(f)]
        best = min(r.priority for r in file_matches) if file_matches else None
        chosen = [r.label for r in file_matches if r.priority == best]
        print(f"  {f} -> {', '.join(chosen) if chosen else '(no match)'}")

    print(f"\nFinal label set: {sorted(labels)}")

File: results/2026-05-06_173435/tasks/12-pr-label-assigner/typescript-bun-sonnet/generated-code/label-assigner.ts
// PR Label Assigner
// Assigns GitHub PR labels based on configurable glob-pattern rules.
// Rules are evaluated in priority order (lower number = higher priority).
// All matching rules contribute their labels; output is deduplicated and sorted.

export interface LabelRule {
  /** Glob pattern to match against changed file paths (e.g., "docs/", ".test.") */
  pattern: string;
  /** Label to apply when the pattern matches */
  label: string;
  /** Lower number = higher priority. Priority only affects evaluation order;
   *  all matching rules always contribute their labels. */
  priority: number;
}

export interface LabelConfig {
  rules: LabelRule[];
}

/**
 * Assigns labels to a PR based on changed file paths and a config.
 * Returns a sorted, deduplicated array of label strings.
 */
export function assignLabels(files: string[], config: LabelConfig): string[] {
  if (files.length === 0 || config.rules.length === 0) {
    return [];
  }

  // Evaluate rules in priority order (lower number first)
  const sortedRules = [...config.rules].sort((a, b) => a.priority - b.priority);
  const labelSet = new Set<string>();

  for (const file of files) {
    for (const rule of sortedRules) {
      const glob = new Bun.Glob(rule.pattern);
      if (glob.match(file)) {
        labelSet.add(rule.label);
      }
    }
  }

  return Array.from(labelSet).sort();
}

// Default configuration covering common project patterns
export const DEFAULT_CONFIG: LabelConfig = {
  rules: [
    { pattern: "docs/**", label: "documentation", priority: 1 },
    { pattern: "*.md", label: "documentation", priority: 1 },
    { pattern: "src/api/**", label: "api", priority: 2 },
    { pattern: "**/*.test.*", label: "tests", priority: 3 },
    { pattern: "**/*.spec.*", label: "tests", priority: 3 },
    { pattern: ".github/**", label: "ci/cd", priority: 4 },
  ],
};

// Demo mode: run with fixture inputs and print parseable results
if (import.meta.main) {
  const fixtures: Array<{ name: string; files: string[] }> = [
    { name: "docs-only", files: ["docs/README.md", "docs/api.md"] },
    { name: "mixed-docs-api", files: ["docs/README.md", "src/api/routes.ts"] },
    { name: "test-files", files: ["src/utils.test.ts", "src/api.test.ts"] },
    { name: "multi-label-single-file", files: ["src/api/routes.test.ts"] },
    { name: "no-match", files: ["random/file.txt", "build/output.bin"] },
    { name: "ci-files", files: [".github/workflows/ci.yml", ".github/dependabot.yml"] },
  ];

  console.log("=== PR Label Assigner Demo ===");
  for (const fixture of fixtures) {
    const labels = assignLabels(fixture.files, DEFAULT_CONFIG);
    const labelStr = labels.length > 0 ? labels.join(",") : "(none)";
    console.log(`[FIXTURE:${fixture.name}] LABELS: ${labelStr}`);
  }
  console.log("=== End Demo ===");
}
